rpcs3/Emu/RSX/RSXThread.h

```diff
@@ -59,56 +55,54 @@ namespace rsx
 			return this->ea[offs >> 20] | (offs & 0xFFFFF);
 		}
 
-		template<bool IsFullLock>
+		template <bool IsFullLock, uint Stride>
 		bool lock(u32 addr, u32 len, cpu_thread* self = nullptr) noexcept
 		{
 			if (len <= 1) return false;
 			const u32 end = addr + len - 1;
 
-			for (u32 block = (addr >> 20); block <= (end >> 20); ++block)
+			bool added_wait = false;
+
+			for (u32 block = addr / 8192; block <= (end / 8192); block += Stride)
 			{
 				auto& mutex_ = rs[block];
 
-				if constexpr (IsFullLock)
+				if (IsFullLock ? !mutex_.try_lock() : !mutex_.try_lock_shared()) [[ likely ]]
```

Comment:
This really isn't likely. It is very unlikely that the millions of r/w operations will actually collide with RSX/SPU writing at the same time.
I do have some worry about the generated code when merging the two branches, I'll have to check manually though. I'll reply to this comment with findings.